contracts/bonding/BondingManager.sol

```diff
@@ -548,6 +560,40 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         emit Bond(_to, currentDelegate, _owner, _amount, del.bondedAmount);
     }
 
+    /**
+     * @notice Checkpoints a delegator state after changes, to be used for historical voting power calculations in
+     * on-chain governor logic.
+     */
+    function checkpointBondingState(
+        address _owner,
+        Delegator storage _delegator,
+        Transcoder storage _transcoder
+    ) internal {
+        // start round refers to the round where the checkpointed stake will be active. The actual `startRound` value
+        // in the delegators doesn't get updated on bond or claim earnings though, so we use currentRound() + 1
+        // which is the only guaranteed round where the currently stored stake will be active.
+        uint256 startRound = roundsManager().currentRound() + 1;
+        bondingCheckpoints().checkpointBondingState(
+            _owner,
+            startRound,
+            _delegator.bondedAmount,
+            _delegator.delegateAddress,
+            _delegator.delegatedAmount,
+            _delegator.lastClaimRound,
+            _transcoder.lastRewardRound
+        );
+    }
+
+    /**
+     * @notice Checkpoints the bonding state for a given account.
+     * @dev This is to allow checkpointing an account that has an inconsistent checkpoint with its current state.
+     * Implemented as a deploy utility to checkpoint the existing state when deploying the BondingCheckpoints contract.
```

Comment:
Questions:

1. Is there a description of how this is supposed to work?
2. Is the intent to call this function for every address? If so:
    a. Do function calls against the BondingManager need to be prevented until this is complete?
    b. What is the cost estimate?

Reply:
I'll answer shortly here for the record, but let's continue the discussion on Discord.

1. I could swear I had included in the spec but apparently not. Need to write this down in some places like LIP and docs.
2. Yeah, to call it for every address that is participating in the protocol, which is already ~4000 from our explorer rn.
   a. They don't! Any action on BondingManager will automatically make a checkpoint as well and has the same end result of calling this function directly.
   b. Around ~200k gas per checkpoint call. The L2 cost seems negligible and something we could pay for, but I don't know much about L1 calldata costs yet.
   
Uncheckpointed accounts used to not be queryable, and after https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3 they just return a 0 voting power until checkpointing happens. If the account hasn't been checkpointed, it will have no voting power. So it is in the best interest of the account to have this checkpoint if they intend to vote, and we can even leave for users themselves to call this instead.

So there are some levels of investment we could make which would be fine, just need proper communication with users:
- We can checkpoint all the participant accounts and leave no actions required from users
- We can checkpoint only the active transcoders (they get their whole voting power from that, dont need each delegator to do it) and leave the task for delegators that want to guarantee their voting power to either call this function or make any bonding change (including `claimRewards`, which is the easiest)
- We checkpoint no account ourselves and have an integration in the explorer to call this function if needed
- We checkpoint no account at all and just let users automatically do it when making changes to BondingManager (`reward()`, `claimEarnings()`, etc)